File: 2.0.2/effects/color_quantization.py
# effects/color_quantization.py
import numpy as np
from PIL import Image

def apply_color_quantization(image, params, selections=None):
    num_colors = int(params['num_colors'])
    dither_amount = float(params['dither_amount'])
    dither_mode = params['dither_mode']
    bit_reduction = int(params['bit_reduction'])

    img_array = np.array(image)

    # Apply bit reduction
    if bit_reduction < 8:
        img_array = (img_array >> (8 - bit_reduction)) << (8 - bit_reduction)

    # Create evenly spaced color levels
    levels = np.linspace(0, 255, num_colors)

    # Apply dithering based on mode
    if dither_amount > 0:
        dither = generate_dither(dither_mode, img_array, dither_amount)
        img_array = np.clip(img_array + dither, 0, 255)

    # Quantize the color channels
    if selections:
        for start, end, channel in selections:
            img_array[:, start:end, channel] = quantize_channel(img_array[:, start:end, channel], levels)
    else:
        img_array = quantize_all_channels(img_array, levels)

    return Image.fromarray(img_array.astype(np.uint8))

def generate_dither(mode, img_array, amount):
    height, width, channels = img_array.shape

    if mode == 'ordered_5x3':
        dither_matrix = np.array([
            [0, 8, 2, 10, 4],
            [12, 4, 14, 6, 1],
            [3, 11, 5, 9, 7]
        ]) / 15 * amount
        dither_matrix = np.tile(dither_matrix, (height // 3 + 1, width // 5 + 1, 1))[:height, :width, np.newaxis]
        dither = np.repeat(dither_matrix, channels, axis=2)

    elif mode == 'ordered_4x1':
        dither_matrix = np.array([[0, 2, 1, 3]]) / 4 * amount
        dither_matrix = np.tile(dither_matrix, (height, width // 4 + 1, 1))[:height, :width, np.newaxis]
        dither = np.repeat(dither_matrix, channels, axis=2)

    elif mode == 'ordered_3x3':
        dither_matrix = np.array([
            [0, 7, 3],
            [6, 5, 2],
            [4, 1, 8]
        ]) / 9 * amount
        dither_matrix = np.tile(dither_matrix, (height // 3 + 1, width // 3 + 1, 1))[:height, :width, np.newaxis]
        dither = np.repeat(dither_matrix, channels, axis=2)

    elif mode == 'ordered_8x8':
        dither_matrix = np.array([
            [0, 48, 12, 60, 3, 51, 15, 63],
            [32, 16, 44, 28, 35, 19, 47, 31],
            [8, 56, 4, 52, 11, 59, 7, 55],
            [40, 24, 36, 20, 43, 27, 39, 23],
            [2, 50, 14, 62, 1, 49, 13, 61],
            [34, 18, 46, 30, 33, 17, 45, 29],
            [10, 58, 6, 54, 9, 57, 5, 53],
            [42, 26, 38, 22, 41, 25, 37, 21]
        ]) / 64 * amount
        dither_matrix = np.tile(dither_matrix, (height // 8 + 1, width // 8 + 1, 1))[:height, :width, np.newaxis]
        dither = np.repeat(dither_matrix, channels, axis=2)

    else:
        # Default to random dithering if no mode is selected
        dither = np.random.uniform(-amount, amount, img_array.shape)

    return dither

def quantize_channel(channel, levels):
    return np.digitize(channel, levels) - 1

def quantize_all_channels(img_array, levels):
    quantized = np.digitize(img_array, levels) - 1
    return levels[quantized]

File: webapp/static/js/effects/color_quantization.js
// webapp/static/js/effects/color_quantization.js
'use strict';

/**
 * Apply color quantization effect to image data.
 * Reduces the number of colors in an image with optional dithering.
 * 
 * @param {ImageData} imageData - The source image data
 * @param {Object} params - Effect parameters
 * @param {number} params.num_colors - Number of colors (2-256)
 * @param {number} params.dither_amount - Dither strength (0-1)
 * @param {string} params.dither_mode - Dithering mode ('none', 'ordered_4x4', 'ordered_8x8', 'random')
 * @returns {ImageData} The processed image data
 */
export function applyColorQuantization(imageData, params) {
    const { width, height, data } = imageData;
    const result = new Uint8ClampedArray(data);
    
    // Parse and clamp parameters
    const numColors = Math.max(2, Math.min(256, Math.floor(params.num_colors || 8)));
    const ditherAmount = Math.max(0, Math.min(1, params.dither_amount || 0.5));
    const ditherMode = params.dither_mode || 'none';
    
    // Create quantization levels
    const levels = [];
    for (let i = 0; i < numColors; i++) {
        levels.push(Math.round((i * 255) / (numColors - 1)));
    }
    
    // Generate dither pattern if needed
    let ditherPattern = null;
    if (ditherAmount > 0 && ditherMode !== 'none') {
        ditherPattern = generateDitherPattern(ditherMode, width, height, ditherAmount);
    }
    
    // Apply quantization
    for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
            const idx = (y * width + x) * 4;
            
            for (let c = 0; c < 3; c++) {
                let value = data[idx + c];
                
                // Apply dither noise
                if (ditherPattern) {
                    const noise = ditherPattern[(y % ditherPattern.length)][(x % ditherPattern[0].length)];
                    value = Math.max(0, Math.min(255, value + noise * 64 * ditherAmount));
                }
                
                // Find closest quantization level
                result[idx + c] = findClosestLevel(value, levels);
            }
            
            // Alpha unchanged
            result[idx + 3] = data[idx + 3];
        }
    }
    
    return new ImageData(result, width, height);
}

/**
 * Find the closest quantization level for a value.
 */
function findClosestLevel(value, levels) {
    let closest = levels[0];
    let minDist = Math.abs(value - closest);
    
    for (let i = 1; i < levels.length; i++) {
        const dist = Math.abs(value - levels[i]);
        if (dist < minDist) {
            minDist = dist;
            closest = levels[i];
        }
    }
    
    return closest;
}

/**
 * Generate dither pattern based on mode.
 */
function generateDitherPattern(mode, width, height, amount) {
    if (mode === 'ordered_4x4') {
        return [
            [0, 8, 2, 10].map(v => (v / 16 - 0.5)),
            [12, 4, 14, 6].map(v => (v / 16 - 0.5)),
            [3, 11, 1, 9].map(v => (v / 16 - 0.5)),
            [15, 7, 13, 5].map(v => (v / 16 - 0.5))
        ];
    } else if (mode === 'ordered_8x8') {
        return [
            [0, 48, 12, 60, 3, 51, 15, 63].map(v => (v / 64 - 0.5)),
            [32, 16, 44, 28, 35, 19, 47, 31].map(v => (v / 64 - 0.5)),
            [8, 56, 4, 52, 11, 59, 7, 55].map(v => (v / 64 - 0.5)),
            [40, 24, 36, 20, 43, 27, 39, 23].map(v => (v / 64 - 0.5)),
            [2, 50, 14, 62, 1, 49, 13, 61].map(v => (v / 64 - 0.5)),
            [34, 18, 46, 30, 33, 17, 45, 29].map(v => (v / 64 - 0.5)),
            [10, 58, 6, 54, 9, 57, 5, 53].map(v => (v / 64 - 0.5)),
            [42, 26, 38, 22, 41, 25, 37, 21].map(v => (v / 64 - 0.5))
        ];
    } else if (mode === 'random') {
        // Generate random noise pattern (small tile that repeats)
        const pattern = [];
        for (let y = 0; y < 8; y++) {
            const row = [];
            for (let x = 0; x < 8; x++) {
                row.push((Math.random() - 0.5));
            }
            pattern.push(row);
        }
        return pattern;
    }
    
    return null;
}
